pennylane_qulacs/qulacs_device.py

```diff
@@ -39,181 +74,240 @@
 
-        expectation = inner_product(bra, self._state)
+        Args:
+            operations (List[pennylane.Operation]): operations to be applied
+        """
 
-        return expectation.real
+        for op in operations:
+            wires = op.wires
+            par = op.parameters
 
-    def probabilities(self):
-        states = itertools.product(range(2), repeat=self.num_wires)
-        probs = np.abs(self.state)**2
+            mapped_operation = self._operation_map[op.name]
+            if op.name.endswith(".inv"):
+                mapped_operation = self._get_inverse_operation(mapped_operation, wires, par)
 
-        return OrderedDict(zip(states, probs))
+            if op.name in ["QubitStateVector", "QubitStateVector.inv"]:
+                input_state = par[0]
+                input_state = _reverse_state(input_state)
 
-    def reset(self):
-        self._state.set_zero_state()
-        self._circuit = QuantumCircuit(self.num_wires)
+                if len(input_state) != 2**len(wires):
+                    raise ValueError("State vector must be of length 2**wires.")
+                if not np.isclose(np.linalg.norm(input_state, 2), 1.0, atol=tolerance):
+                    raise ValueError("Sum of amplitudes-squared does not equal one.")
+                # call qulac"s state initialization
+                self._state.load(input_state)
+
+            elif op.name in ["BasisState", "BasisState.inv"]:
+                # translate from PennyLane to Qulacs wire order
+                bits = par[0][::-1]
+                n_basis_state = len(bits)
+
+                if not set(bits).issubset({0, 1}):
+                    raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
+                if n_basis_state != len(wires):
+                    raise ValueError("BasisState parameter and wires must be of equal length.")
+
+                basis_state = 0
+                for bit in bits:
+                    basis_state = (basis_state << 1) | bit
+
+                # call qulac's basis state initialization
+                self._state.set_computational_basis(basis_state)
+
+            elif op.name in ["QubitUnitary", "QubitUnitary.inv", "Hermitian", "Hermitian.inv"]:
+                if len(par[0]) != 2 ** len(wires):
+                    raise ValueError("Unitary matrix must be of shape (2**wires, 2**wires).")
+
+                if op.name in ["QubitUnitary.inv", "Hermitian.inv"]:
+                    par[0] = par[0].conj().T
+
+                # reverse wires (could also change par[0])
+                unitary_gate = gate.DenseMatrix(wires[::-1], par[0])
+                self._circuit.add_gate(unitary_gate)
+                unitary_gate.update_quantum_state(self._state)
+
+            elif op.name in ["Rot", "Rot.inv"]:
+                if len(wires) != 1:
+                    raise ValueError("Rotation gate can only be applied on a single wire.")
```

Comment:
This check is not needed, it should be taken care of by the part of PennyLane core that is device independent:
```python
import pennylane as qml

dev = qml.device('default.qubit', wires=2)

@qml.qnode(dev)
def circuit():
    qml.Rot(0,0,0, wires=[0,1])
    return qml.expval(qml.PauliZ(0))

circuit()
```
```
~/xanadu/pennylane/pennylane/operation.py in __init__(self, wires, do_queue, *params)
    377             raise ValueError(
    378                 "{}: wrong number of wires. "
--> 379                 "{} wires given, {} expected.".format(self.name, len(self._wires), self.num_wires)
    380             )
    381 

ValueError: Rot: wrong number of wires. 2 wires given, 1 expected.
```

Each `Operation` has a `num_wires` attribute and there is a check placed using that.